File: utils/get_frame.cpp
#include <opencv2/opencv.hpp>
#include <array>
#include <iostream>

#include "framedata.hpp"


void get_frame(mjModel* model, mjData* data, mjvOption* opt, mjvScene* scene, mjrContext* context, mjvCamera* cam, 
    FrameData& frameData, int loc_x, int loc_y, int width, int height) {
    // 1. Create a rectangular viewport
    mjrRect offscreen_viewport = {loc_x, loc_y, width, height};

    // 2. Update scene for off-screen camera
    mjv_updateScene(model, data, opt, NULL, cam, mjCAT_ALL, scene);

    // 3. Render the scene in the offscreen buffer
    unsigned char* rgb = (unsigned char*)malloc(height * width * 3 * sizeof(unsigned char));
    float* depth = (float*)malloc(height * width * 1 * sizeof(float));
    unsigned char* depth8 = (unsigned char*)malloc(height * width * 3 * sizeof(unsigned char));
    float* depthm = (float*)malloc(height * width * 1 * sizeof(float));

    mjr_render(offscreen_viewport, scene, context);

    // 4. Read the pixels
    mjr_readPixels(rgb, depth, offscreen_viewport, context);

    // convert to meters
    float extent = model->stat.extent;
    float near = model->vis.map.znear * extent;
    float far = model->vis.map.zfar * extent;
    for (int i=0; i< height * width; ++i) {
    depthm[i] = near / (1.0f - depth[i] * (1.0f - near/far));
    }

    // Normalize depth to fit within 8-bit range
    float min_depth = FLT_MAX;
    float max_depth = -FLT_MAX;
    for (int i = 0; i < height * width; i++) {
    if (depth[i] < min_depth) min_depth = depth[i];
    if (depth[i] > max_depth) max_depth = depth[i];
    }

    float range = max_depth - min_depth;
    for (int i = 0; i < height * width; i++) {
    depth8[3 * i] = depth8[3 * i + 1] = depth8[3 * i + 2] = (unsigned char)((depth[i] - min_depth) / range * 255.0f);
    }

    // mjr_drawPixels(depth8, NULL, offscreen_viewport, context);

    // Package the data
    frameData.rgb = rgb;
    frameData.depth = depth;
    frameData.depthm = depthm;
    frameData.depth8 = depth8;
    frameData.viewport = offscreen_viewport;
    frameData.width = width;
    frameData.height = height;
}

File: utils/get_frame.py
import mujoco as mj
import numpy as np


def get_frame(model, data, opt, scene, context, camera_name, loc_x, loc_y, width=640, height=480):
    # Bottom Placement
    # bottom left: loc_x = 0, loc_y = 0
    # bottom middle: loc_x = 0.5*(viewport_width - width), loc_y = 0
    # bottom right: loc_x = viewport_width - width, loc_y = 0
    # Middle Placement
    # middle left: loc_x = 0, loc_y = 0.5*(viewport_width - width)
    # middle: loc_x = 0.5*(viewport_width - width), loc_y = 0.5*(viewport_width - width)
    # middle right: loc_x = viewport_width - width, loc_y = 0.5*(viewport_width - width)
    # Top Placement
    # top left: loc_x = 0, loc_y = viewport_height - height
    # top middle: loc_x = 0.5*(viewport_width - width), loc_y = viewport_height - height
    # top right: loc_x = viewport_width - width, loc_y = viewport_height - height
    
    height = int(height)
    width = int(width)
    # Adding an inset window from a different perspective
    # https://github.com/google-deepmind/mujoco/issues/744#issuecomment-1442221178
    # 1. Create a rectangular viewport in the upper right corner for example.
    offscreen_viewport = mj.MjrRect(int(loc_x), int(loc_y), width, height)
    
    # 2. Specify a different camera view by updating the scene with mjv_updateScene.
    # Set the camera to the specified view
    camera_id = mj.mj_name2id(model, mj.mjtObj.mjOBJ_CAMERA, camera_name)
    offscreen_cam = mj.MjvCamera()
    offscreen_cam.type = mj.mjtCamera.mjCAMERA_FIXED
    offscreen_cam.fixedcamid = camera_id

    # Update scene for the off-screen camera
    mj.mjv_updateScene(model, data, opt, None, offscreen_cam, mj.mjtCatBit.mjCAT_ALL.value, scene)
    
    # 3.Render the scene in the offscreen buffer with mjr_render.
    frame = np.zeros((height*width*3, 1), dtype=np.uint8)  # Placeholder for pixel data
    mj.mjr_render(offscreen_viewport, scene, context)
    
    # 4. Read the pixels with mjr_readPixels.
    mj.mjr_readPixels(frame, None, offscreen_viewport, context)
    
    # 5. Call mjr_drawPixels using the rectangular viewport you created in step 1.
    # glClear(GL_DEPTH_BUFFER_BIT)  # allows rendering over geometries
    # mj.mjr_drawPixels(frame, None, offscreen_viewport, context)
    
    return frame, offscreen_viewport